src/persist/src/s3.rs

```diff
@@ -920,17 +872,39 @@ impl Iterator for MultipartChunkIter {
 }
 
 /// A helper for tracking the minimum of a set of Durations.
-#[derive(Debug, Default)]
-struct MinElapsed(Option<Duration>);
+#[derive(Debug)]
+struct MinElapsed {
+    min: AtomicU64,
+    alert_factor: u64,
+}
+
+impl Default for MinElapsed {
+    fn default() -> Self {
+        MinElapsed {
+            min: AtomicU64::new(u64::MAX),
+            alert_factor: 8,
+        }
+    }
+}
 
 impl MinElapsed {
-    /// Records a new Duration and returns the minimum seen so far.
-    fn observe(&mut self, x: Duration) -> Duration {
-        let min = self.0.get_or_insert(x);
-        if x < *min {
-            *min = x;
+    fn observe(&self, x: Duration, msg: &'static str) {
+        let nanos = x.as_nanos();
+        let nanos = u64::try_from(nanos).unwrap_or(u64::MAX);
+
+        // Possibly set a new minimum.
+        self.min.fetch_min(nanos, atomic::Ordering::SeqCst);
+
+        // Trace if our provided duration was much larger than our minimum.
+        let new_min = self.min.load(atomic::Ordering::Relaxed);
+        let current_duration = Duration::from_nanos(nanos);
```

Comment:
may as well use x here? you don't get the u64::max truncation, but seems fine

Reply:
Done!